File: frontend/src/app-editorial/day-bounds.ts
/** Local-calendar-day helpers shared by the "today" meal log and the 30-day
 * meal history, so both agree on where one day ends and the next begins. */

export function startOfDay(date: Date): Date {
  const d = new Date(date)
  d.setHours(0, 0, 0, 0)
  return d
}

export function addDays(date: Date, days: number): Date {
  const d = new Date(date)
  d.setDate(d.getDate() + days)
  return d
}

/** Local YYYY-MM-DD key for grouping timestamps into calendar days. */
export function dayKey(date: Date): string {
  const y = date.getFullYear()
  const m = String(date.getMonth() + 1).padStart(2, "0")
  const d = String(date.getDate()).padStart(2, "0")
  return `${y}-${m}-${d}`
}

/** Milliseconds until the next local midnight, for scheduling a refresh. */
export function msUntilNextMidnight(from: Date = new Date()): number {
  return addDays(startOfDay(from), 1).getTime() - from.getTime()
}

File: ios/Fuel/Core/Logic/DayBounds.swift
import Foundation

// PURE Foundation port of frontend/src/app-editorial/day-bounds.ts. Local-
// calendar-day helpers shared by the Today log, 30-day History, streaks and the
// week chart, so they all agree where one day ends and the next begins.
//
// The TS uses the device's local calendar (`Date.setHours`, `getDay`,
// `getFullYear`…); we mirror that with `Calendar.current`. A calendar is
// injectable so tests are deterministic regardless of the machine's time zone.
enum DayBounds {
  /// Local midnight at the start of `date`'s calendar day.
  static func startOfDay(_ date: Date, calendar: Calendar = .current) -> Date {
    calendar.startOfDay(for: date)
  }

  /// `date` shifted by whole calendar days (DST-safe, like JS `setDate`).
  static func addDays(_ date: Date, _ days: Int, calendar: Calendar = .current) -> Date {
    calendar.date(byAdding: .day, value: days, to: date) ?? date
  }

  /// Local YYYY-MM-DD key for grouping timestamps into calendar days.
  static func dayKey(_ date: Date, calendar: Calendar = .current) -> String {
    let c = calendar.dateComponents([.year, .month, .day], from: date)
    return String(format: "%04d-%02d-%02d", c.year ?? 0, c.month ?? 0, c.day ?? 0)
  }

  /// Seconds until the next local midnight, for scheduling a refresh.
  static func intervalUntilNextMidnight(from: Date = Date(), calendar: Calendar = .current) -> TimeInterval {
    addDays(startOfDay(from, calendar: calendar), 1, calendar: calendar).timeIntervalSince(from)
  }
}
